recipes-app/iot2050-firmware-update/files/iot2050-firmware-update.py

```diff
@@ -0,0 +1,193 @@
+        except FileNotFoundError:
+            print("Open /dev/mem Failed")
+            sys.exit(1)
+
+        mem = mmap.mmap(f, mmap.PAGESIZE, mmap.MAP_SHARED, mmap.PROT_READ, offset=base_addr)
+        mem.seek(base_addr_offset)
+        data = []
+        data.append(struct.unpack('I', mem.read(4))[0])
+        device_id = hex(data[0])
+        current_id = int(device_id, 16) >> 11
+        if current_id not in cpu_device_id:
+            FirmwareUpdateException("Upgrade is not supported for the board")
+
+    def update_firmware(self, filename):
+        """Update Firmware"""
+        mtd_num = 0
+        img_pos = 0
+        skip_bytes = 0
+
+        print("===================================================")
+        print("IOT2050 firmware update started - DO NOT INTERRUPT!")
+        print("===================================================")
+
+        while True:
+            mtd_sys_path = "/sys/bus/platform/devices/47040000.spi/mtd/mtd{}".format(mtd_num)
+            mtd_name_path = "{}/name".format(mtd_sys_path)
+            mtd_size_path = "{}/size".format(mtd_sys_path)
+            mtd_dev_path = "/dev/mtd{}".format(mtd_num)
+
+            mtd_size = self.get_path_type_value(mtd_size_path)
+            mtd_size = int(mtd_size)
+
+            mtd_name = self.get_path_type_value(mtd_name_path)
+            env_list = ["ospi.env","ospi.env.backup"]
+
+            if not self.erase_env and mtd_name.strip() in env_list:
+                skip_bytes = skip_bytes+mtd_size
+                mtd_num = mtd_num+1
+                print("Skipping update %-20s" % mtd_name)
+                continue
+            else:
+                print("Updating %-20s" % mtd_name.strip(),end="")
+
+            if mtd_name.strip() == "ospi.rootfs":
+                print("\n\nCompleted. Please reboot the device\n")
+                return
+            try:
+                with open(mtd_dev_path, 'rb') as mtd:
+                    mtd_contents = mtd.read()
+            except FileNotFoundError:
+                print("Open %s failed" % mtd_dev_path)
+                sys.exit(1)
+
+            with open(filename, 'rb') as firmware:
+                firmware.seek(img_pos+skip_bytes)
+                firmware_contents = firmware.read(mtd_size)
+
+            if not mtd_contents == firmware_contents:
+                print("U")
+                self.flash_erase(mtd_dev_path, 0, mtd_size)
+                with open(mtd_dev_path, 'wb') as mtd_unity:
+                    mtd_unity.write(firmware_contents)
+            else:
+                print(".")
+            img_pos = img_pos+mtd_size
+            mtd_num = (mtd_num+1)
+            print("\n")
+
+
+def main(argv):
+    parser = OptionParser()
+    parser.add_option('-f', '--force', action='store_true', dest='force', default=False, help="Force update the firmware,ignore the error")
+    parser.add_option('-e', '--erase', action='store_true', dest='erase', default=False, help="Erase the Environment")
```

Comment:
Just realized: --erase must be default on unless we add a method to ensure that the vars u-boot normally writes are updated.

Reply:
If do that, we can not retain the environment that user sets and changes.

I found it can not parse the VARS "default_environment" from .bin file.
It could combine the ELF file(uboot) to get some symbol info and get the VARS from the decompile .bin to do the upgrade.  Or just provide the symbol info of the "default_environment" along with upgrade bin file.

Do you some suggestions how to retain the  environment  for both of the user and the upgrade Firmware Simultaneously

Reply:
As with many low-level updates, keeping user settings is an add-on, not a must-have. The higher value is maintaining consistency, and that implies in its basic form overwriting the env.

The problem is that even the default env, eventually available via a new package (feature under development elsewhere), is not enough because we transfer dynamic information into the env, see https://github.com/siemens/u-boot/blob/dfdc967909de39a8abfaf02551f2a3fd8d24ac8a/board/siemens/iot2050/board.c#L56. By default, this must be in sync after an update, thus requires to reset the env. Well, at least to clear `board_uuid` in order to trigger rewriting of the other vars on first boot. But clearing `board_uuid` would also overwrite potential customizations of MAC addresses or fdtfile or who knows.

And, finally, the risk that keeping the old env will break booting with the new firmware version is simply too high, e.g. when switching from the current downstream U-Boot to upstream. Or when switching between current open boot and locked-down secure boot. So we shouldn't try this by default, maybe not even at all. Experts editing the env can also easily reapply changes to the new version.